blog/2022/05-10-resoto-graph-traversal/index.md

```diff
@@ -0,0 +1,119 @@
+
+```bash
+> search is(digitalocean_droplet) and id=289061882 <-- | count kind
+// highlight-start
+​digitalocean_region: 1
+​digitalocean_image: 1
+​digitalocean_vpc: 1
+​digitalocean_kubernetes_cluster: 1
+​digitalocean_load_balancer: 1
+​total matched: 5
+​total unmatched: 0
+// highlight-end
+```
+
+The droplet belongs to a region inside a [VPC](https://docs.digitalocean.com/products/networking/vpc/). It is the target of a load balancer and part of a kubernetes cluster. Let's find out which other instances exist in the same VPC:
+
+```bash
+> search is(digitalocean_droplet) and id=289061882
+  <-- is(digitalocean_vpc) --> is(digitalocean_droplet)
+// highlight-start
+​id=289061882, name=dev-1, age=2mo7d, account=10225075, region=Amsterdam 3
+​id=289061881, name=dev-3, age=2mo7d, account=10225075, region=Amsterdam 3
+​id=289061880, name=dev-2, age=2mo7d, account=10225075, region=Amsterdam 3
+// highlight-end
+```
+
+We again start at our droplet and walk the graph inbound to the VPC. From there, we walk one step outbound to find all droplets attached to the same VPC. There are three droplets in this VPC, and one of them is the one we were starting from. It is possible to chain any number of graph traversals. We could have used the same technique to find all the droplets in the Kubernetes cluster or the ones behind the load balancer.
+
+Since this droplet is attached to a load balancer, we want to find out the public IP address. If multiple load balancers were attached to the same droplet, we would get the list of all IP addresses.
+
+```bash
+> search is(digitalocean_droplet) and id=289061882 <-- is(load_balancer) |
+  format {public_ip_address}
+// highlight-start
+​137.121.131.102
+// highlight-end
+```
+
+Resoto allows for walking not only one step but multiple steps. You can define a specific number of steps to walk and a range of steps. You can even specify the range as unlimited, which would traverse the graph to the root or leaf, depending on the direction of the walk. The following search will select all DigitalOcean VPCs and then walk the graph outbound, returning all nodes that can be reached starting from the VPC up to two steps. This list would never include snapshots since snapshots can not be reached in two steps starting from the VPC.
+
+```bash
+> search is(digitalocean_vpc) -[0:2]-> | count kind
+// highlight-start
+​digitalocean_app: 1
+​digitalocean_database: 1
+​digitalocean_load_balancer: 2
+​digitalocean_kubernetes_cluster: 2
+​digitalocean_vpc: 2
+​digitalocean_volume: 6
+​digitalocean_droplet: 6
+​total matched: 20
+​total unmatched: 0
+// highlight-end
+```
+
+If we can walk outbound with `-->` and inbound with `<--`, there has to be an option to walk in both directions. The combination of the two approaches is expressed with `<-->`. It is possible to get the list of all reachable resources starting from a specific node. We use the above droplet and ask for all attached resources.
```

Comment:
I think the reader would benefit from  introducing the three concepts at the top of the post, and then repeating them here. 

At the beginning of the post, I think you want to have a little explanatory paragraph that explain  the graph, plus inbound / outbound and both directions. Could be done in bullets.